KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,652 @@
+
+<!-- All KIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The KIP must explain how the author proposes to deal with these incompatibilities. KIP submissions without a sufficient backwards compatibility treatise may be rejected outright. The authors should answer the question: "Does this KIP require a hard fork?" -->
+
+- The foundation should deploy new TreasuryRebalance contract to record the token redistribution.
+- To rebalance the funds and redistribute in a consistent manner the foundation should burn the designated funds before re-distribution.
+- This does not affect the backward compatibility as this a newly dpeloyed contract
+
+## Implementation
+
+<!--The implementations must be completed before any KIP is given status "Final", but it need not be completed before the KIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
+
+#### Example implementation
+
+```solidity
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity ^0.8.0;
+
+import "./Ownable.sol";
+
+/**
+ * @title Smart contract to record the rebalance of treasury funds.
+ * This contract is to mainly record the addresses which holds the treasury funds
+ * before and after rebalancing. It facilates approval and redistributing to new addresses.
+ * Core will execute the re-distribution reading from this contract.
+ */
+contract TreasuryRebalance is Ownable {
+    /**
+     *  Enums to track the status of the contract
+     */
+    enum Status {
+        Initialized,
+        Registered,
+        Approved,
+        Finalized
+    }
+
+    /**
+     * Retired struct to store the details of retired and approver addresses
+     */
+    struct Retired {
+        address retired;
+        address[] approvers;
+    }
+
+    /**
+     * Newbie struct to store reciever and amount
+     */
+    struct Newbie {
+        address newbie;
+        uint256 amount;
+    }
+
+    /**
+     * Storage
+     */
+    Retired[] public retirees; //array of retired structs
+    Newbie[] public newbies; //array of newbie structs
+    Status public status; //current status of the contract
+    uint256 public rebalanceBlockNumber; //Block number of the execution of rebalancing
+    string public memo; //result of the treasury fund rebalance
+
+    /**
+     * Events logs
+     */
+    event DeployContract(
+        Status status,
+        uint256 rebalanceBlockNumber,
+        uint256 deployedBlockNumber
+    );
+    event RegisterRetired(address retired, address[] approvers);
+    event RemoveRetired(address retired, uint256 retiredCount);
+    event RegisterNewbie(address newbie, uint256 fundAllocation);
+    event RemoveNewbie(address newbie, uint256 newbieCount);
+    event GetState(bool success, bytes result);
```

Comment:
Why do we need to have GetState event? Isn't this a read operation?

Reply:
Yes but this is read from another contract, so it would be good to log the results. Isn't ?

Reply:
@toniya-klaytn But it is little bit weird to use the name `GetState`. If you want to log the result of calling getstate() function, we can use other names like `RetiredAddressAdminStatus`.

Reply:
updated to `RetiredAddrAdminStatus` instead of `RetiredAddressAdminStatus` as too long